File: crypto_bot/data/indicator_engine.py
from __future__ import annotations

from data.data_quality import QualityInput, resolve_quality


def _closes(candles: list[dict]) -> list[float]:
    return [float(row["close"]) for row in candles if row.get("close") is not None]


def _highs(candles: list[dict]) -> list[float]:
    return [float(row["high"]) for row in candles if row.get("high") is not None]


def _lows(candles: list[dict]) -> list[float]:
    return [float(row["low"]) for row in candles if row.get("low") is not None]


def _volumes(candles: list[dict]) -> list[float]:
    return [float(row.get("volume") or 0.0) for row in candles]


def calculate_atr(candles: list[dict], period: int = 14) -> float | None:
    if len(candles) < max(int(period) + 1, 2):
        return None
    highs = _highs(candles)
    lows = _lows(candles)
    closes = _closes(candles)
    if len(highs) != len(candles) or len(lows) != len(candles) or len(closes) != len(candles):
        return None

    trs: list[float] = []
    for i in range(1, len(candles)):
        h = highs[i]
        l = lows[i]
        prev_close = closes[i - 1]
        tr = max(h - l, abs(h - prev_close), abs(l - prev_close))
        trs.append(tr)
    if len(trs) < int(period):
        return None
    last = trs[-int(period):]
    return sum(last) / len(last)


def calculate_rsi(candles: list[dict], period: int = 14) -> float | None:
    closes = _closes(candles)
    if len(closes) < int(period) + 1:
        return None
    gains: list[float] = []
    losses: list[float] = []
    for i in range(1, len(closes)):
        delta = closes[i] - closes[i - 1]
        gains.append(max(delta, 0.0))
        losses.append(max(-delta, 0.0))
    window_gains = gains[-int(period):]
    window_losses = losses[-int(period):]
    avg_gain = sum(window_gains) / len(window_gains)
    avg_loss = sum(window_losses) / len(window_losses)
    if avg_loss == 0:
        return 100.0
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def calculate_return_volatility(candles: list[dict], period: int = 20) -> float | None:
    closes = _closes(candles)
    if len(closes) < int(period) + 1:
        return None
    returns: list[float] = []
    for i in range(1, len(closes)):
        prev = closes[i - 1]
        curr = closes[i]
        if prev <= 0:
            continue
        returns.append((curr - prev) / prev)
    if len(returns) < int(period):
        return None
    data = returns[-int(period):]
    mean = sum(data) / len(data)
    variance = sum((x - mean) ** 2 for x in data) / len(data)
    return variance ** 0.5


def calculate_bundle(
    candles: list[dict],
    *,
    min_points: int = 30,
    stale: bool = False,
    supported: bool = True,
) -> dict:
    closes = _closes(candles)
    highs = _highs(candles)
    lows = _lows(candles)
    volumes = _volumes(candles)
    count = len(candles)

    if count == 0 or not closes:
        quality = resolve_quality(
            QualityInput(
                sample_count=count,
                min_required=min_points,
                stale=stale,
                supported=supported,
                reason="no_candles",
            )
        )
        return {
            "atr": None,
            "rsi": None,
            "return_volatility": None,
            "candle_range_pct": None,
            "momentum": None,
            "recent_high": None,
            "recent_low": None,
            "compression_score": None,
            "expansion_score": None,
            "vwap": None,
            "data_quality": quality,
        }

    atr = calculate_atr(candles, period=14)
    rsi = calculate_rsi(candles, period=14)
    return_vol = calculate_return_volatility(candles, period=20)
    recent_high = max(highs[-20:]) if highs else None
    recent_low = min(lows[-20:]) if lows else None
    latest_close = closes[-1]
    prev_close = closes[-2] if len(closes) >= 2 else latest_close
    momentum = ((latest_close - prev_close) / prev_close) if prev_close > 0 else None
    candle_range_pct = (
        ((highs[-1] - lows[-1]) / latest_close)
        if highs and lows and latest_close > 0
        else None
    )
    typical = [
        (float(row["high"]) + float(row["low"]) + float(row["close"])) / 3.0
        for row in candles
        if row.get("high") is not None and row.get("low") is not None and row.get("close") is not None
    ]
    volume_sum = sum(volumes)
    vwap = (
        (sum(tp * vol for tp, vol in zip(typical, volumes)) / volume_sum)
        if volume_sum > 0 and len(typical) == len(volumes)
        else None
    )
    norm_vol = return_vol if return_vol is not None else 0.0
    compression_score = max(0.0, min(100.0, (1.0 - min(norm_vol / 0.02, 1.0)) * 100.0))
    expansion_score = max(0.0, min(100.0, min(norm_vol / 0.02, 1.0) * 100.0))

    quality_reason = "ok"
    if count < min_points:
        quality_reason = "insufficient_history"
    elif any(item is None for item in (atr, rsi, return_vol)):
        quality_reason = "partial_indicator_coverage"
    quality = resolve_quality(
        QualityInput(
            sample_count=count,
            min_required=min_points,
            stale=stale,
            supported=supported,
            reason=quality_reason,
        )
    )
    return {
        "atr": atr,
        "rsi": rsi,
        "return_volatility": return_vol,
        "candle_range_pct": candle_range_pct,
        "momentum": momentum,
        "recent_high": recent_high,
        "recent_low": recent_low,
        "compression_score": compression_score,
        "expansion_score": expansion_score,
        "vwap": vwap,
        "data_quality": quality,
    }

File: web-ui/src/app/lib/indicatorEngine.mjs
function clamp(value, min, max) {
  return Math.max(min, Math.min(max, value));
}

function toCandles(pricePoints, bucketSeconds = 60) {
  if (!Array.isArray(pricePoints) || pricePoints.length === 0) {
    return [];
  }
  const bucket = Math.max(1, Math.trunc(bucketSeconds));
  const buckets = new Map();
  for (const row of pricePoints) {
    const ts = Number(row?.tsEpoch ?? 0);
    const price = Number(row?.price ?? 0);
    if (!Number.isFinite(ts) || ts <= 0 || !Number.isFinite(price) || price <= 0) {
      continue;
    }
    const key = Math.floor(ts / bucket) * bucket;
    const existing = buckets.get(key);
    if (!existing) {
      buckets.set(key, {
        tsEpoch: key,
        open: price,
        high: price,
        low: price,
        close: price,
        volume: 1,
      });
      continue;
    }
    existing.high = Math.max(existing.high, price);
    existing.low = Math.min(existing.low, price);
    existing.close = price;
    existing.volume += 1;
  }
  return Array.from(buckets.values()).sort((a, b) => a.tsEpoch - b.tsEpoch);
}

function atr(candles, period = 14) {
  if (!Array.isArray(candles) || candles.length < period + 1) {
    return null;
  }
  const tr = [];
  for (let i = 1; i < candles.length; i += 1) {
    const h = candles[i].high;
    const l = candles[i].low;
    const prevClose = candles[i - 1].close;
    tr.push(Math.max(h - l, Math.abs(h - prevClose), Math.abs(l - prevClose)));
  }
  if (tr.length < period) {
    return null;
  }
  const slice = tr.slice(-period);
  return slice.reduce((sum, value) => sum + value, 0) / slice.length;
}

function rsi(candles, period = 14) {
  if (!Array.isArray(candles) || candles.length < period + 1) {
    return null;
  }
  let gains = 0;
  let losses = 0;
  for (let i = candles.length - period; i < candles.length; i += 1) {
    const prev = candles[i - 1].close;
    const curr = candles[i].close;
    const delta = curr - prev;
    if (delta > 0) {
      gains += delta;
    } else {
      losses += Math.abs(delta);
    }
  }
  const avgGain = gains / period;
  const avgLoss = losses / period;
  if (avgLoss <= 0) {
    return 100;
  }
  const rs = avgGain / avgLoss;
  return 100 - (100 / (1 + rs));
}

function returnVolatility(candles, period = 20) {
  if (!Array.isArray(candles) || candles.length < period + 1) {
    return null;
  }
  const values = [];
  for (let i = candles.length - period; i < candles.length; i += 1) {
    const prev = candles[i - 1].close;
    const curr = candles[i].close;
    if (prev <= 0) {
      continue;
    }
    values.push((curr - prev) / prev);
  }
  if (values.length < 3) {
    return null;
  }
  const mean = values.reduce((s, x) => s + x, 0) / values.length;
  const variance = values.reduce((s, x) => s + ((x - mean) ** 2), 0) / values.length;
  return Math.sqrt(Math.max(variance, 0));
}

function buildIndicatorBundle({
  symbol,
  pricePoints,
  wallClockEpoch = Date.now() / 1000,
}) {
  const candles = toCandles(pricePoints, 60);
  const latest = candles[candles.length - 1] ?? null;
  const earliest = candles[0] ?? null;
  const spanMinutes = (latest && earliest) ? Math.max((latest.tsEpoch - earliest.tsEpoch) / 60, 0) : 0;
  const atrValue = atr(candles, 14);
  const rsiValue = rsi(candles, 14);
  const volValue = returnVolatility(candles, 20);
  const recent = candles.slice(-20);
  const recentHigh = recent.length > 0 ? Math.max(...recent.map((row) => row.high)) : null;
  const recentLow = recent.length > 0 ? Math.min(...recent.map((row) => row.low)) : null;
  const latestClose = latest?.close ?? null;
  const prevClose = candles.length >= 2 ? candles[candles.length - 2].close : null;
  const momentum = (latestClose !== null && prevClose !== null && prevClose > 0)
    ? ((latestClose - prevClose) / prevClose)
    : null;
  const rangePct = (latest && latest.close > 0)
    ? ((latest.high - latest.low) / latest.close)
    : null;
  const volNorm = volValue === null ? null : clamp((volValue / 0.02) * 100, 0, 100);
  const compressionScore = volNorm === null ? null : (100 - volNorm);
  const expansionScore = volNorm;
  const vwap = candles.length > 0
    ? candles.reduce((sum, row) => sum + row.close * row.volume, 0)
      / Math.max(candles.reduce((sum, row) => sum + row.volume, 0), 1e-9)
    : null;

  const ageMinutes = latest ? Math.max((wallClockEpoch - latest.tsEpoch) / 60, 0) : null;
  let quality = "GOOD";
  let reason = "ok";
  if (candles.length < 30) {
    quality = candles.length > 0 ? "PARTIAL" : "INSUFFICIENT";
    reason = candles.length > 0 ? "insufficient_samples" : "no_samples";
  }
  if (ageMinutes !== null && ageMinutes > 20) {
    quality = "STALE";
    reason = "stale_history";
  }

  return {
    symbol,
    atr: atrValue,
    rsi: rsiValue,
    returnVolatility: volValue,
    candleRangePct: rangePct,
    momentum,
    recentHigh,
    recentLow,
    compressionScore,
    expansionScore,
    vwap,
    observedPointCount: candles.length,
    observedHistorySpanMinutes: Number(spanMinutes.toFixed(3)),
    dataQuality: {
      status: quality,
      reason,
      sampleCounts: {
        observed: candles.length,
        required: 30,
      },
      lastUpdateTs: latest ? new Date(latest.tsEpoch * 1000).toISOString() : null,
    },
  };
}

export {
  toCandles,
  buildIndicatorBundle,
};
